Fix notify inactive collators failures at the end of a round

manuelmauro opened the pull request:
### What does it do?
Addresses issue #3100 for which "notify_inactive_collator" only works at the beginning of a round.

### What important points reviewers should know?

### Is there something left for follow-up PRs?

### What alternative implementations were considered?

### Are there relevant PRs or issues in other repositories (Substrate, Polkadot, Frontier, Cumulus)?

### What value does it bring to the blockchain users?

crystalin commented:
For optimization purposes (knowing that Storage/PoV is usually the limiting factor), shouldn't we store those who didn't have stake. Instead of `HadStake` we could have `WasInactive` ? (What do you think @RomarQ ?)

RomarQ commented on pallets/parachain-staking/src/lib.rs:
```diff
@@ -503,6 +503,7 @@ pub mod pallet {
 		}
 		fn on_finalize(_n: BlockNumberFor<T>) {
 			Self::award_points_to_block_author();
+			Self::cleanup_stake_info();
```
We need to account for this extra weight in `on_initialize`.

RomarQ commented:
> For optimization purposes (knowing that Storage/PoV is usually the limiting factor), shouldn't we store those who didn't have stake. Instead of `HadStake` we could have `WasInactive` ? (What do you think @RomarQ ?)

Yes, having a `WasInactive` storage map is preferable:

```rs
pub type WasInactive<T: Config> =
		StorageDoubleMap<_, Twox64Concat, RoundIndex, Twox64Concat, T::AccountId, (), OptionQuery>;
```

When starting a new round, we would check the inactivity like the example below, where `stake.is_some() && pts.is_zero()` would mean that the collator is inactive. We would check it against every candidate.

https://github.com/moonbeam-foundation/moonbeam/blob/fef382138efaa065a7c88130a6b1bcd2e753289a/pallets/parachain-staking/src/lib.rs#L1427-L1430

We would still clear the storage after `MaxOfflineRounds`

RomarQ commented on pallets/parachain-staking/src/lib.rs:
```diff
@@ -2345,6 +2354,21 @@ pub mod pallet {
+				return;
+			}
+
+			let _ = <HadStake<T>>::iter_prefix(now - minimum_rounds_required)
+				.drain()
+				.next();
+		}
```
If we properly estimate the weights in `on_initialize`, we could remove all entries from `WasInactive` for the round here.

manuelmauro commented on pallets/parachain-staking/src/mock.rs:
```diff
@@ -277,7 +277,10 @@ impl ExtBuilder {
+fn roll_one_block(finalize_parachain_staking: bool) -> BlockNumber {
+	if finalize_parachain_staking {
+		ParachainStaking::on_finalize(System::block_number());
+	}
```
Addressed in https://github.com/moonbeam-foundation/moonbeam/pull/3128/commits/325137dea41cfcc6ad04c4c2eaff7140cb3fad61

Please double check this changes in particular:
https://github.com/moonbeam-foundation/moonbeam/pull/3128/commits/325137dea41cfcc6ad04c4c2eaff7140cb3fad61#diff-a2fcd9e2ec1e588e9fcfed674d8d0058713d685aaba3ed374ba248d4c279e02fL6778-R6781

**Note on the changes introduced in this commit**: the function previously used to mimic the distribution of points and to obviate to the missing call to `ParachainStaking::on_finalize` directly modifies the storage in a similar way to what `on_finalize` does (although matching logic is not enforced).

```rust
// Same storage changes as ParachainStaking::on_finalize
pub(crate) fn set_author(round: BlockNumber, acc: u64, pts: u32) {
	<Points<Test>>::mutate(round, |p| *p += pts);
	<AwardedPts<Test>>::mutate(round, acc, |p| *p += pts);
}
```

With the call to `on_finalize` within block rolling, and the concurrent use of this function, rounds may assign a variable amount of points making check on rewards amounts unreliable.

IMO a better approach would be to introduce utility functions as:

```rust
/// Sets the block author and advances n blocks.
pub(crate) fn author_blocks(n: BlockNumber, acc: u64) -> BlockNumber {
	set_block_author(acc);
	roll_blocks(n)
}

/// Sets the block author and advances blocks until the beginning of the specified round.
pub(crate) fn author_to_round_begin(round: BlockNumber, acc: u64) -> BlockNumber {
	set_block_author(acc);
	roll_to_round_begin(round)
}
```

That would guarantee a higher fidelity of the mocking framework.

manuelmauro commented:
> For optimization purposes (knowing that Storage/PoV is usually the limiting factor), shouldn't we store those who didn't have stake. Instead of `HadStake` we could have `WasInactive` ? (What do you think @RomarQ ?)

Addressed comment, as described by @RomarQ,  in https://github.com/moonbeam-foundation/moonbeam/pull/3128/commits/b4f4c005915b8f7a2d7355428531360e203367a1.
Now fixing weights.

RomarQ commented on pallets/parachain-staking/src/lib.rs:
```diff
@@ -489,20 +489,23 @@ pub mod pallet {
 					total_balance: total_staked,
 				});
+				// record inactive collators
+				Self::mark_collators_as_inactive(round.current);
```
```suggestion
				weight = weight.saturating_add(Self::mark_collators_as_inactive(round.current));
```

manuelmauro replied:
Addressed in https://github.com/moonbeam-foundation/moonbeam/pull/3128/commits/74cbae1cc254831ba1235df4b3af97c46f965075 👍

RomarQ commented on pallets/parachain-staking/src/lib.rs:
```diff
@@ -2333,18 +2335,46 @@ pub mod pallet {
+			// We don't need to saturate here because the genesis round is 1.
+			let prev = cur - 1;
+
+			for (account, _) in <AtStake<T>>::iter_prefix(prev) {
```
This currently does not impact the implementation, but it was an extrinsic, a possible improvement would be to wrap the loop inside `if <WasInactive<T>>::iter_prefix(prev).next().is_none()`.

RomarQ commented on pallets/parachain-staking/src/lib.rs:
```diff
@@ -489,20 +489,23 @@ pub mod pallet {
 				weight = weight.saturating_add(Self::handle_delayed_payouts(round.current));
 			}
 
 			// add on_finalize weight
```
We are ignoring the proof size for both `award_points_to_block_author` and `cleanup_inactive_collator_info`, but should be fine since we over-estimate above.

RomarQ reviewed (commented):
After putting some thought into the `mark_collators_as_inactive` weights, the over-estimation is probably not necessary, we could re-use the iterator in `select_top_candidates` to update the `WasInactive` storage.

manuelmauro commented:
> After putting some thought into the `mark_collators_as_inactive` weights, the over-estimation is probably not necessary, instead we should add a `num_of_collators` argument to the `mark_collators_as_inactive` benchmark.

Nice! Addressed in https://github.com/moonbeam-foundation/moonbeam/pull/3128/commits/b66ec1fe033543b63f1ed02957d099c69d098cc8
